package/MDAnalysis/topology/TOPParser.py

```diff
@@ -298,12 +307,36 @@ def next_getter():
         attrs['atomids'] = Atomids(np.arange(n_atoms) + 1)
         attrs['resids'] = Resids(np.arange(n_res) + 1)
         attrs['resnums'] = Resnums(np.arange(n_res) + 1)
-        attrs['segids'] = Segids(np.array(['SYSTEM'], dtype=object))
 
-        top = Topology(n_atoms, n_res, 1,
-                       attrs=list(attrs.values()),
-                       atom_resindex=residx,
-                       residue_segindex=None)
+        try:
+            if len(attrs["segids"]) != n_res:
+                msg = (
+                    f"Number of residues ({n_res}) does not match number of "
+                    f"%RESIDUE_CHAINID ({len(attrs['segids'])}). Skipping section."
+                )
+                logger.warning(msg)
+                warnings.warn(msg)
+                raise ValueError
+
+            segidx, (segids,) = change_squash((attrs["segids"],), (attrs["segids"],))
+            chainids = [attrs["segids"][r] for r in residx]
+
+            attrs["segids"] = Segids(segids)
+            attrs["ChainIDs"] = ChainIDs(chainids)
+            n_segs = len(segids)
+        except (KeyError, ValueError):
```

Comment:
If I'm reading this, the default option is that the optional attribute (which isn't going to be present in most PARM7 files), is the option you hope will happen and then throw an error in all other cases?

I'm not very keen on this, since it means that we have to eat up the cost of throwing an error for 99% of the time. Is there a better way to reformat this so we don't have to deal with the error throwing / catching cost in most cases?

Reply:
Done. Using ifs now.